src/main/java/teammates/logic/api/Logic.java

```diff
@@ -785,16 +786,19 @@ public static boolean isCourseArchived(CourseAttributes course, InstructorAttrib
     /**
      * Preconditions: <br>
      * * All parameters are non-null.
+     * <br>
+     * @param alreadyRetrievedData The previously obtained query results to be re-used
      * 
      * @return Details of courses the student is in. CourseData objects
      *         returned contain details of feedback sessions too (except the ones
      *         still AWAITING).
      */
-    public List<CourseDetailsBundle> getCourseDetailsListForStudent(String googleId)
-            throws EntityDoesNotExistException {
+    public List<CourseDetailsBundle> getCourseDetailsListForStudent(String googleId,
+                                              Map<String, StudentAttributes> alreadyRetrievedData)
+                                     throws EntityDoesNotExistException {
         
         Assumption.assertNotNull(ERROR_NULL_PARAMETER, googleId);
-        return coursesLogic.getCourseDetailsListForStudent(googleId);
+        return coursesLogic.getCourseDetailsListForStudent(googleId, alreadyRetrievedData);
```

Comment:
I'll see if I can carry that out. If I manage to, it can be reused in so many other operations.